src/backer/cardaning.py

```diff
@@ -59,44 +65,104 @@ def __init__(self, name='backer', hab=None, ks=None):
 
+    def updateTip(self, tipHeight):        
+        self.tipHeight = tipHeight
+
     def publishEvent(self, event: bytes):
         self.pending_kel = event + self.pending_kel
 
-    def flushQueue(self):
+    def submitKelTx(self, kel):        
         try:
-            if self.pending_kel:
+            if kel:
                 # Build transaction
                 builder = pycardano.TransactionBuilder(self.context)
                 builder.add_input_address(self.spending_addr)
                 builder.add_output(pycardano.TransactionOutput(self.spending_addr, pycardano.Value.from_primitive([TRANSACTION_AMOUNT])))
-                self.pending_kel = bytes(self.pending_kel)
+                kel_data = bytes(kel)
                 # Chunk size
                 # bytearrays is not accept
-                value = [self.pending_kel[i:i + 64] for i in range(0, len(self.pending_kel), 64)]
+                value = [kel_data[i:i + 64] for i in range(0, len(kel_data), 64)]
 
                 # Metadata. accept int key type
                 builder.auxiliary_data = pycardano.AuxiliaryData(pycardano.Metadata({1: value}))
 
                 signed_tx = builder.build_and_sign([self.payment_signing_key],
                                                 change_address=self.spending_addr,
                                                 merge_change=True)
+                
                 # Submit transaction
-                self.context.submit_tx(signed_tx.to_cbor())  
-                self.pending_kel = bytearray()
-        except Exception as e:    
-            logger.critical(f"Error: {e}")
+                self.context.submit_tx(signed_tx.to_cbor())
+                transId = str(signed_tx.transaction_body.inputs[0].transaction_id)
+                trans = {
+                    "id": transId,
+                    "kel": kel.decode('utf-8'),
+                    "publish_time": help.toIso8601()
+                }
+                self.keldbConfirming.pin(keys=transId, val=json.dumps(trans).encode('utf-8'))
+        except Exception as e:
+            logger.critical(f"Cannot submit transaction: {e}")
+
+    def emptyPendingKEL(self):
+        self.pending_kel = bytearray()
+
+    def flushQueue(self):
+        if self.pending_kel is not None and self.pending_kel != b"":
+            self.submitKelTx(self.pending_kel)        
+            self.pending_kel = bytearray()
+
+    def slotToTime(self, slot):
+        timeStamp = SHELLY_UNIX + slot
+        return datetime.fromtimestamp(timeStamp)
+
+    def getConfirmingTrans(self, transId):
+        return self.keldbConfirming.get(transId)
+
+    def confirmTrans(self):
+        try:
+            for keys, item in self.keldbConfirming.getItemIter():
+                if item is None:
+                    continue
+                
+                item = json.loads(item)
+                (blockSlot, blockHeight) = (item["block_slot"], item["block_height"]) if 'block_slot' in item.keys() else (0, 0)
```

Comment:
Hmm, is it ever possible for these not to be defined and end up with `(0, 0)` (the else block)?

Reply:
or is this before it appears on chain at all?

Reply:
> or is this before it appears on chain at all?
Yes.
We have doer to execute the pending confirmation, so if the trans is not on chain there is not block_slot and block_height at that time

Reply:
you're right, I will add more condition for this: tipHeight and blockHeight must be large than 0